ntcore/ds_client.py

```diff
@@ -1,21 +1,133 @@
-# validated: 0000-00-00 DS 000000000 cpp/DsClient.cpp cpp/DsClient.h
-'''----------------------------------------------------------------------------'''
-''' Copyright (c) FIRST 2017. All Rights Reserved.                             '''
-''' Open Source Software - may be modified and shared by FRC teams. The code   '''
-''' must be accompanied by the FIRST BSD license file in the root directory of '''
-''' the project.                                                               '''
-'''----------------------------------------------------------------------------'''
+# validated: 2017-12-09 DV 5ab20bb27c97 cpp/DsClient.cpp cpp/DsClient.h
+#----------------------------------------------------------------------------
+# Copyright (c) FIRST 2017. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+#----------------------------------------------------------------------------
+
+import json
+import threading
+
+from .support.compat import Condition
+from .tcpsockets.tcp_connector import TcpConnector
 
 import logging
 logger = logging.getLogger('nt')
 
+
 class DsClient(object):
     
     def __init__(self, dispatcher, verbose=False):
-        pass
-    
+        self.m_dispatcher = dispatcher
+        self.verbose = verbose
+
+        self.m_active = False
+        self.m_owner = None  # type: threading.Thread
+
+        self.m_mutex = threading.Lock()
+        self.m_cond = Condition(self.m_mutex)
+
+        self.m_port = None  # type: int
+        self.m_stream = None
+
     def start(self, port):
-        logging.warn("Dashboard client mode not currently implemented")
-    
+        with self.m_mutex:
+            self.m_port = port
+            if not self.m_active:
+                self.m_active = True
+                self.m_owner = threading.Thread(target=self._thread_main,
+                                                name='nt-dsclient-thread')
+                self.m_owner.daemon = True
+                self.m_owner.start()
```

Comment:
I'm guessing the original implementation uses SafeThread? I have an implementation in ntcore/support.

Reply:
I had a look at it before, and I wasn't sure if the things upstream uses are actually available in our implementation of it. I might try it again though.